README
"moon-phase-video" notes                               Revised: 120709
----------------------------------------------------------------------

1. Overview.

1.1. "moon-phase-video"  is  a set of Perl 5 programs that generates a
short  video clip showing  one month in the life of the Moon. Both GIF
animation and AVI versions are produced.

To see the  video clip, go to the "examples" directory which is stored
next  to this README file and view the  GIF file(s) there using a  web
browser or "viewnior" (a program linked at a later point). Or view the
AVI file(s) using a  standard multimedia player  such as  "mplayer" or
"vlc".

There are three source files (two Perl5 programs and one data file):

(a) moonphasepic  - Perl5 program that generates a single video frame.

(b) moonphase2avi - Perl5 program that calls "moonphasepic" repeatedly
to generate individual frames, then combines the frames to produce ei-
ther an animated GIF or an AVI file.

(c) albedo_simp750.jpg - Base image used by "moonphasepic".

1.2. "moon-phase-video" was created by OldCoder:

      Site:    http://oldcoder.org/
      Mail:    oldcoder@yahoo.com
      GitHub:  https://github.com/OldCoder/
      Repo:    https://github.com/OldCoder/moon-phase-video.git

The  license used for the current version is Creative Commons Attribu-
tion NonCommercial  ShareAlike  3.0  Unported.  The data file "albedo_
simp750.jpg"  is an exception;  the file in question is believed to be
in the public domain.

----------------------------------------------------------------------

2. Requirements.

Requirements are  moderately  stringent.  To use this package,  you'll
need:

Linux, Perl 5.8 or above, ImageMagick (or "imagemagick") 6.3 or above,
"mencoder"(often bundled with "mplayer"), POV-Ray (or "povray") 3.6.1,
"gifsicle" 1.48 or above,  "intergif" 6.15 or above,  and release 0.60
or above of a CPAN module named Astro::MoonPhase.

Perl comes preinstalled in most Linux distros.

"imagemagick", "mencoder",  and "povray" are  generally available from
Linux distro repositories.

"gifsicle" and  "intergif"  may be  available from distro repositories
or,  for some distros,  you may need to build these  two programs from
source code.

Astro::MoonPhase is  available  from CPAN.  For more information about
CPAN, see the associated link below.

Related links:

      gifsicle:  http://freecode.com/projects/gifsicle/

      intergif:  http://freecode.com/projects/intergif/

      mencoder:  Same as for "mplayer"

      mplayer:   http://www.mplayerhq.hu/

      povray:    http://povray.org/

      viewnior:  http://freecode.com/projects/viewnior/

      vlc:       http://www.videolan.org/vlc/

      CPAN:      http://www.cpan.org/

----------------------------------------------------------------------

3. Installation.

No "build" is needed.

To install  the  package, simply copy the three source files listed in
part 1 to wherever you'd like to store them.  The files should go into
the same directory and  the user should have  read, write, and execute
permissions for the directory.

----------------------------------------------------------------------

4. Usage.

4.1. Displaying documentation.

To use  either of the Perl5 scripts provided, go to the directory that
holds them and execute a command of the form:

      perl FILENAME.pl

Alternatively,  remove the ".pl" filename extensions and set file per-
missions to 755 octal. If this is done, commands of the following form
should work:

      FILENAME          # If "." is     in PATH
or  ./FILENAME          # If "." is not in PATH

If no command-line arguments are specified, program documentation will
be printed in each case.

4.2. "moonphase2avi" summary.

Usage:  moonphase2avi --avi         # Make high-quality AVI
        moonphase2avi --gif         # Make animated GIF instead
        moonphase2avi --needs       # List software requirements
        moonphase2avi --license     # Show license information

"--avi" creates  an AVI file  called "moonphase.avi" that displays one
month in the life of  the Moon.  The AVI file is MPEG4-encoded. You'll
need  write  access to  the  current directory and  about 8 MB of disk
space. However, the AVI file is only about 210 KB in size.

"--gif" is similar, but  it  creates a low-resolution animated GIF in-
stead of an AVI.

For software requirements or license information, specify "--needs" or
"--license".

4.3. "moonphasepic" summary.

Usage:  moonphasepic now   --output=foo.png
        moonphasepic 0.52  --output=foo.png

        --info=foo.txt    Create an informational text file
        --width=360       Change image width

This program creates a PNG image file. The image shows the Moon with a
phase shadow added.

"--output" specifies the image-file name (or pathname).  If "--output"
isn't used, the image is saved to  "moonphase.png" in  the user's home
directory.

If "now" is specified, the phase is based on the current date.  A num-
ber between  zero and  one (inclusive) may be  specified instead.  The
number specifies a point  in the lunar cycle  (for example: zero = New
Moon, 0.25 = 1st quarter, 0.50 = Full Moon, and 0.75 = 3rd quarter).

"--info=foo.txt" writes a short description of the image to the speci-
fied text file. If "--info" isn't used, no text file is created.

To change the image width, use "--width" as shown above.


CHANGELOG
"moon-phase-video" ChangeLog
----------------------------

2012-07-09 rev. 120709:

First GitHub release.


source/moonphasepic.pl
#!/usr/bin/env perl
# moonphasepic - Generates a Moon picture with phase shadow added
# License:  Creative Commons Attribution-NonCommercial-ShareAlike 3.0
# Revision: 120709

#---------------------------------------------------------------------
#                           important note
#---------------------------------------------------------------------

# This program is  provided on an  AS IS basis with ABSOLUTELY NO WAR-
# RANTY. The entire risk as to the quality and performance of the pro-
# gram is with you. Should the program prove defective, you assume the
# cost of all necessary  servicing, repair or correction.  In no event
# will any of the developers, or any other party,  be liable to anyone
# for damages arising out of  use of the program,  or inability to use
# the program.

#---------------------------------------------------------------------
#                              overview
#---------------------------------------------------------------------

                                # 'END...' may be single-quoted here
my $DOCUMENTATION = << 'END_OF_DOCUMENTATION';
__META_NAME__ rev. __META_REVISION__
- Generates a Moon picture with phase shadow

Recommended: Remove the ".pl"  filename extension and set file permis-
sions to 755 octal.

Usage:  __META_NAME__ now   --output=foo.png
        __META_NAME__ 0.52  --output=foo.png

        --info=foo.txt    Create an informational text file
        --width=360       Change image width

This program creates a PNG image file. "--output" specifies the image-
file name (or pathname).  If "--output" isn't used, the image is saved
to "moonphase.png" in the user's home directory.

If "now" is specified, the phase is based on the current date.  A num-
ber between  zero and  one (inclusive) may be  specified instead.  The
number specifies a point  in the lunar cycle  (for example: zero = New
Moon, 0.25 = 1st quarter, 0.50 = Full Moon, and 0.75 = 3rd quarter).

"--info=foo.txt" writes a short description of the image to the speci-
fied text file.  If "--info" isn't used, no text file is  created.  To
change the image width, use "--width" as shown above.
END_OF_DOCUMENTATION

#---------------------------------------------------------------------
#                          base module setup
#---------------------------------------------------------------------

require 5.8.1;
use strict;
use Carp;
use warnings;
                                # Trap warnings
$SIG {__WARN__} = sub { die @_; };

#---------------------------------------------------------------------
#                           basic constants
#---------------------------------------------------------------------

use constant ZERO  => 0;        # Zero
use constant ONE   => 1;        # One

use constant FALSE => 0;        # Boolean FALSE
use constant TRUE  => 1;        # Boolean TRUE

                                # Approx. value of "pi"
my $PI = 3.14159265358979323846264338327950288419716939937510;

my $DEGREES_PER_HALF   = 180;   # Number of degrees per half-circle
my $DEGREES_PER_CIRCLE = 360;   # Number of degrees per circle
                                # Number of radians per degree
my $RADIANS_PER_DEGREE = (2 * $PI) / $DEGREES_PER_CIRCLE;

#---------------------------------------------------------------------
#                         additional modules
#---------------------------------------------------------------------

use Cwd;                        # Provided with Perl
use Math::Trig;                 # Ditto

my $HaveCPAN = FALSE;           # Flag:  Have CPAN module
                                # Try to load CPAN module
eval 'use Astro::MoonPhase; $HaveCPAN = TRUE;';
                                # See if module was loaded
die "Error: The CPAN module Astro::MoonPhase is needed\n"
    unless $HaveCPAN;

#---------------------------------------------------------------------
#                         program parameters
#---------------------------------------------------------------------

my $PROGNAME  = 'moonphasepic'; # Program name (should be one word)
my $REVISION  = '120709';       # Revision string
                                # Name of base-image file
my $IMAGE_MAP = "albedo_simp750.jpg";

my $LIGHT_DISTANCE  = 5000;     # Should be from 100 to 5000
my $LIGHT_INTENSITY = 1.70;     # Should be from 1.0 to 2.0

#---------------------------------------------------------------------

# $DEFAULT_WIDTH and $DEFAULT_HEIGHT  specify the default output width
# and height, respectively (in pixels).  The factory settings are  400
# and 300. Note: The associated aspect ratio should be 4/3.

my $DEFAULT_WIDTH  = 400;
my $DEFAULT_HEIGHT = 300;

#---------------------------------------------------------------------

#  The "$POVTMP..." files are used  primarily to  work around restric-
# tions imposed by POV-Ray. Specifically, POV-Ray complains when path-
# names are used for files. It doesn't seem to like  symbolic links to
# absolute paths  either.  As a  work-around,  we move files around so
# that POV-Ray doesn't need to see paths.

# Note: This problem is a POV-Ray configuration issue.  Theoretically,
# users could fix the problem by modifying their POV-Ray configuration
# files, but we'd prefer to avoid this.

#  $POVTMPDIR          = Absolute pathname for temporary directory
my $POVTMPDIR          = "/var/tmp";

#  $POVTMPNAME_IMGFILE = Name of image output file without path
#  $POVTMPNAME_MAPFILE = Name of image map    file without path
#  $POVTMPNAME_SRCCODE = Name of POV-Ray code file without path

my $POVTMPNAME_IMGFILE = "moonphase-out-$>-$$.png";
my $POVTMPNAME_MAPFILE = "moonphase-map-$>-$$.jpg";
my $POVTMPNAME_SRCCODE = "moonphase-src-$>-$$.pov";

#  $POVTMPPATH_IMGFILE = Name of image output file with path
#  $POVTMPPATH_MAPFILE = Name of image map    file with path
#  $POVTMPPATH_SRCCODE = Name of POV-Ray code file with path

my $POVTMPPATH_IMGFILE = "$POVTMPDIR/$POVTMPNAME_IMGFILE";
my $POVTMPPATH_MAPFILE = "$POVTMPDIR/$POVTMPNAME_MAPFILE";
my $POVTMPPATH_SRCCODE = "$POVTMPDIR/$POVTMPNAME_SRCCODE";

#---------------------------------------------------------------------
#                        POV-Ray code template
#---------------------------------------------------------------------

# The following  block of code is written in  POV-Ray's scene descrip-
# tion language.

my $POV_CODE = << 'END_OF_POV_CODE';
#include "colors.inc"
#include "functions.inc"

light_source {
    <__META_X__, 0, __META_Z__>
    color <__META_LIGHT_COLOR__>
}

camera {
    location <0, 0, -2.5>
    look_at  <0, 0, 0>
}

#local pg_fn = function {
    pigment {
        image_map {
            jpeg "__META_IMAGE_MAP__"
            map_type 1
            interpolate 2
        }
    }
}

isosurface {
    function {
        f_sphere (x, y, z, 1)
    }
    max_gradient 1.810
    texture {
        pigment {
            function { pg_fn (x, y, z).gray }
        }

        finish {
            ambient rgb <0, 0, 0.24>
            diffuse 1.0
            specular 0.1
            roughness 0.05
        }
    }

    rotate -90 * y
}
END_OF_POV_CODE

#---------------------------------------------------------------------
#                           misc. routines
#---------------------------------------------------------------------

# "UsageError" prints the program's "usage" text and exits.

sub UsageError
{
    $DOCUMENTATION =~ s@^\s+@@s;
    $DOCUMENTATION =~ s@(__META_REVISION__)\n(-)@$1 $2@;
    $DOCUMENTATION =~ s@__META_NAME__@$PROGNAME@g;
    $DOCUMENTATION =~ s@__META_REVISION__@$REVISION@g;
    $DOCUMENTATION =~ s@\s*\z@\n@s;

    print "\n", $DOCUMENTATION, "\n";
    exit ONE;
}

#---------------------------------------------------------------------
#                            main routine
#---------------------------------------------------------------------

sub Main
{
    my $angle;                  # An angle (expressed in radians)
    my $str;                    # Scratch
    my $time;                   # UNIX time value

    my $info;                   # Information related to image
    my $ofname;                 # Image -file name (or pathname)
    my $txtfile;                # Info  -file name (or pathname)

    my $frac;                   # Point  in  lunar cycle  (ranges from
                                # zero to one, inclusive)

    my $x;                      # X-coordinate of light source
    my $z;                      # Z-coordinate of light source
                                # Output width and height (in pixels)
    my $OutWidth  = $DEFAULT_WIDTH;
    my $OutHeight = $DEFAULT_HEIGHT;

#---------------------------------------------------------------------
# Initial setup.

                                # Note: STDERR must be set first here
    select STDERR; $| = ONE;    # Force STDERR flush on write
    select STDOUT; $| = ONE;    # Force STDOUT flush on write

#---------------------------------------------------------------------
# Process the command line.

    for my $arg (@ARGV)         # Process all  arguments
    {                           # Process next argument
        if ($arg =~ m@^now\z@i)
        {
            &UsageError() if defined $frac;
            $time = time;
            $frac = phase ($time);
            $info = 'time ' . $time;
        }
        elsif (($arg =~ m@^\d+\.?\d*\z@) && ($arg <= 1.0))
        {
            &UsageError() if defined $frac;
            $frac =  $arg;
            $frac =~ s@\.\z@@;
            $info =  'frac ' . $frac;
        }
        elsif ($arg =~ m@^-+info=(\S+)\z@)
        {
            $txtfile =  $1;
            $txtfile =~ s@^["']+@@;
            $txtfile =~ s@["']+\z@@;
        }
        elsif ($arg =~ m@^-+out(|put)=(\S+)\z@)
        {
            $ofname =  $2;
            $ofname =~ s@^["']+@@;
            $ofname =~ s@["']+\z@@;
        }
        elsif (($arg =~ m@^-+width=(\d+)\z@) &&
            ($1 >= 48) && ($1 <= 2048))
        {
            $OutWidth  = $1;
            $OutHeight = int (((3 * $OutWidth) / 4) + 0.5);
        }
        else
        {
            &UsageError();
        }
    }

    &UsageError() unless defined $frac;

#---------------------------------------------------------------------
# Check external programs.

    my @progs = qw (povray);    # Build list of programs needed

    for my $prog (@progs)       # Check all  programs
    {                           # Check next program
        $str =  `which "$prog" 2>&1`;
        $str =  "" unless defined $str;
        $str =~ s@\n.*\z@@s;
                                # Error if program is missing
        die "Error: Operation needs the program $prog\n"
            unless (-f $str) && (-x $str);
    }

#---------------------------------------------------------------------
# Set default output-file name.

    if (!defined ($ofname) || !length ($ofname))
    {
        $str    = $ENV {HOME};
        $str    = '/tmp' unless defined ($str) && length ($str);
        $ofname = "$str/moonphase.png";
    }

#---------------------------------------------------------------------
# Safety measure.

    for my $ref_path (\$ofname, \$txtfile)
    {
        my $xfname = $$ref_path;
        next unless defined $xfname;

        unlink $xfname;         # Remove symbolic link (if necessary)
                                # Try to create the file
        open (OFD, ">$xfname") ||
            die "Error: Can't create file: $!\n$xfname\n";

        print OFD "Test\n";
        close (OFD) ||
            die "Error: Can't write to file: $!\n$xfname\n";

                                # Try to remove the file
        if (!unlink ($xfname) || -e $xfname || -l $xfname)
        {
            die "Error: Can't remove file: $!\n$xfname\n";
        }
    }

#---------------------------------------------------------------------
# Handle required computations.

                                # Point in  lunar cycle,  expressed as
                                # an angle (in degrees)
    my $deg360 = $frac * $DEGREES_PER_CIRCLE;

                                # Safety measure
    $deg360 = $DEGREES_PER_CIRCLE if $deg360 > $DEGREES_PER_CIRCLE;

    my $xs  = ONE;              # Sign of X-coordinate (-1 or +1)
    my $zs  = ONE;              # Sign of Z-coordinate (-1 or +1)

                                # Determine related parameters
    if (($frac >= 0.00) && ($frac < 0.25))
    {                           # 1st quarter
        $angle =  $deg360 * $RADIANS_PER_DEGREE;
        $info  .= "\n1st quarter and waxing\n";
    }
    elsif (($frac >= 0.25) && ($frac <= 0.50))
    {                           # 2nd quarter
        $angle =  ($DEGREES_PER_HALF - $deg360)
                 * $RADIANS_PER_DEGREE;
        $info  .= "\n2nd quarter and waxing\n";
        $zs    =  -1;
    }
    elsif (($frac > 0.50) && ($frac < 0.75))
    {                           # 3rd quarter
        $angle =  ($deg360 - $DEGREES_PER_HALF)
                           * $RADIANS_PER_DEGREE;
        $info  .= "\n3rd quarter and waning\n";
        $xs    =  -1;
        $zs    =  -1;
    }
    else
    {                           # 4th quarter
        $angle =  ($DEGREES_PER_CIRCLE - $deg360)
                 * $RADIANS_PER_DEGREE;
        $info  .= "\n4th quarter and waning\n";
        $xs    =  -1;
    }
                                # Compute X and Z coordinates
    $x = $xs * $LIGHT_DISTANCE * abs (sin ($angle));
    $z = $zs * $LIGHT_DISTANCE * abs (cos ($angle));

#---------------------------------------------------------------------
# Generate a POV-Ray source file.

    my $L = $LIGHT_INTENSITY;

    $str =  $POV_CODE;
    $str =~ s@__META_IMAGE_MAP__@$POVTMPNAME_MAPFILE@g;
    $str =~ s@__META_LIGHT_COLOR__@$L, $L, $L@g;
    $str =~ s@__META_X__@$x@g;
    $str =~ s@__META_Z__@$z@g;

    my $TF = 'temporary file';
    open (OFD, ">$POVTMPPATH_SRCCODE") ||
        die "Error: Can't create $TF: $!\n$POVTMPPATH_SRCCODE\n";
    print OFD $str;
    close (OFD) ||
        die "Error: Can't write to $TF: $!\n$POVTMPPATH_SRCCODE\n";

#---------------------------------------------------------------------
# Run POV-Ray.

# Note: POV-Ray makes some  kludges related to  temporary files neces-
# sary.  For more information,  see the comments preceding the defini-
# tions of $POVTMPDIR near the start of this file.

    my $CWD = getcwd();         # Current working directory
                                # Try to create temporary directory
                                # If  this fails,  it'll  be caught by
                                # the "chdir" command below
    system "/bin/mkdir -p $POVTMPDIR";

                                # Make temporary copy of base image
    system "/bin/cp $IMAGE_MAP $POVTMPPATH_MAPFILE";

                                # Go to temporary directory
    chdir ($POVTMPDIR) ||
        die "Error: Can't go to directory: $!\n$POVTMPDIR\n";

                                # Build appropriate "povray" command
    $str = << "END";
povray +I$POVTMPNAME_SRCCODE +O$POVTMPNAME_IMGFILE
+W$OutWidth +H$OutHeight -D
END
    $str =~ s@\s+@ @gs;
    $str =~ s@\s+\z@@s;
    system $str;                # Execute the "povray" command

    unlink $POVTMPPATH_MAPFILE; # Delete temporary files
    unlink $POVTMPPATH_SRCCODE;
                                # Return to original directory
    chdir ($CWD) ||
        die "Error: Can't go to directory: $!\n$CWD\n";

                                # Move output image to final location
    system "/bin/mv $POVTMPPATH_IMGFILE $ofname";

#---------------------------------------------------------------------
# Generate image-info file (if requested).

    if (defined ($txtfile) && open (OFD, ">$txtfile"))
    {
        if ($info =~ m@^frac (\d+\.?\d*)\n(\S.+?)\s*\z@s)
        {
            my ($p1, $p2) = ($1, $2);
            $p1 *= 100;

            $info = << "END";
This is the Moon ${p1}\% of the way through its cycle. Zero is equal
to New Moon and 50\% is equal to Full Moon. At this point, the Moon
is in its $p2
END
        }
        elsif ($info =~ m@^time (\d+)\n(\S.+?)\s*\z@s)
        {
            my ($p1, $p2) = ($1, $2);
            $str = localtime ($p1);
            $str =~ s@ \d\d:\d\d:\d\d ([A-Z]\w* +|)@ @;
            $str =~ s@\s+@ @g;
            $str =~ s@\s+\z@@s;

            $info = << "END";
This is the Moon as of $str. At this point, the Moon is
in its $p2
END
        }
        else
        {
            $info = 'moonphasepic: Internal error';
        }

        $info =~ s@\s+\z@@s;
        $info =~ s@\.*\z@.@;

        if (($frac >= 0.85) || ($frac <= 0.15))
        {
            $info .= "\n\n";
            $info .= << "END";
Note: The Moon is invisible because it's close to the start of the
lunar cycle.
END
        }

        $info =~ s@\s*\z@\n@;
        print OFD $info;
        close OFD;
    }

#---------------------------------------------------------------------
# Wrap it up.

    if (-f $ofname)
    {
        print "Created $ofname";
        print " and $txtfile" if defined $txtfile;
        print "\n";
    }
    else
    {
        die "Error: Operation failed for unknown reasons\n";
    }

    undef;
}

#---------------------------------------------------------------------
#                            main program
#---------------------------------------------------------------------

&Main();                        # Call the main routine
exit ZERO;                      # Normal exit


source/moonphase2avi.pl
#!/usr/bin/env perl
# moonphase2avi.pl - Creates a moon-phase AVI or an animated GIF
# License:  Creative Commons Attribution-NonCommercial-ShareAlike 3.0
# Revision: 120709

#---------------------------------------------------------------------
#                           important note
#---------------------------------------------------------------------

# This program is  provided on an  AS IS basis with ABSOLUTELY NO WAR-
# RANTY. The entire risk as to the quality and performance of the pro-
# gram is with you. Should the program prove defective, you assume the
# cost of all necessary  servicing, repair or correction.  In no event
# will any of the developers, or any other party,  be liable to anyone
# for damages arising out of  use of the program,  or inability to use
# the program.

#---------------------------------------------------------------------
#                              overview
#---------------------------------------------------------------------

                                # 'END...' may be single-quoted here
my $DOCUMENTATION = << 'END_OF_DOCUMENTATION';
__META_NAME__ rev. __META_REVISION__
- Creates a moon-phase video clip

Recommended: Remove the ".pl"  filename extension and set file permis-
sions to 755 octal.

Usage:  __META_NAME__ --avi         # Make high-quality AVI
        __META_NAME__ --gif         # Make animated GIF instead
        __META_NAME__ --needs       # List software requirements
        __META_NAME__ --license     # Show license information

"--avi" creates  an  AVI file  named "moonphase.avi" that displays one
month in the life of  the Moon.  The AVI file is MPEG4-encoded. You'll
need  write  access to  the  current directory and  about 8 MB of disk
space. However, the AVI file is only about 210 KB in size.

"--gif" is similar, but it creates a low-resolution  animated  GIF in-
stead of an AVI.

For software requirements or license information, specify "--needs" or
"--license".
END_OF_DOCUMENTATION

#---------------------------------------------------------------------
#                          base module setup
#---------------------------------------------------------------------

require 5.8.1;
use strict;
use Carp;
use warnings;
                                # Trap warnings
$SIG {__WARN__} = sub { die @_; };

#---------------------------------------------------------------------
#                           basic constants
#---------------------------------------------------------------------

use constant ZERO  => 0;        # Zero
use constant ONE   => 1;        # One
use constant TWO   => 2;        # Two

use constant FALSE => 0;        # Boolean FALSE
use constant TRUE  => 1;        # Boolean TRUE

#---------------------------------------------------------------------
#                         program parameters
#---------------------------------------------------------------------

my $PROGNAME = 'moonphase2avi'; # Program name (should be one word)
my $REVISION = '120709';        # Revision string

#---------------------------------------------------------------------

# AVI-file parameters:

my $AVI_FPS       =  20;        # Frames per second (10 to 30)
my $AVI_NUMFRAMES = 200;        # No. of frames (100 to 300)
                                # Output    -file name
my $AVI_OFNAME    = 'moonphase.avi';
                                # Temporary -file name
my $AVI_TFNAME    = "temp$$.tmp";
my $AVI_WIDTH     = 320;        # Image width (in pixels)

#---------------------------------------------------------------------

# GIF-file parameters:

my $GIF_FPS       =  10;        # Frames per second (10 to 30)
my $GIF_NUMFRAMES = 150;        # No. of frames (100 to 300)
                                # Output    -file name
my $GIF_OFNAME    = 'moonphase.gif';
                                # Temporary -file name
my $GIF_TFNAME    = "temp$$.gif";
my $GIF_WIDTH     = 120;        # Image width (in pixels)

my $IGNUMCOLORS   =  28;        # Max no. of colors in final GIF file

#---------------------------------------------------------------------
#                            main routine
#---------------------------------------------------------------------

sub Main
{
    my $cmd;                    # Shell-level command string
    my $str;                    # Scratch

    my $cmderr  = FALSE;        # Flag: Command-line error
    my $is_avi  = TRUE;         # Flag: Make AVI (as opposed to GIF)
    my $run     = FALSE;        # Flag: Proceed with main operation

    my $license = FALSE;        # Flag: Saw "--license" switch
    my $needs   = FALSE;        # Flag: Saw "--need(s)" switch

#---------------------------------------------------------------------
# Initial setup.

                                # Note: STDERR must be set first here
    select STDERR; $| = ONE;    # Force STDERR flush on write
    select STDOUT; $| = ONE;    # Force STDOUT flush on write

#---------------------------------------------------------------------
# Process the command line.

    for my $arg (@ARGV)         # Process all  arguments
    {                           # Process next argument
        if ($arg =~ s@^-+(make|)avi\w*\z@@i)
            { $is_avi = TRUE  ; $run = TRUE; }

        if ($arg =~ s@^-+(make|)gif\w*\z@@i)
            { $is_avi = FALSE ; $run = TRUE; }

        $license  = TRUE  if $arg =~ s@^-+lic\w*\z@@i;
        $needs    = TRUE  if $arg =~ s@^-+need\w*\z@@i;
        $needs    = TRUE  if $arg =~ s@^-+require\w*\z@@i;
        $run      = TRUE  if $arg =~ s@^-+run\w*\z@@i;
        $cmderr   = TRUE  if length $arg;
    }

#---------------------------------------------------------------------
# Handle "--license" and/or "--need(s)" switches.

    if ($license || $needs)
    {
        print << 'END';

Requirements: This program requires "moonphasepic" (a related program)
and "povray"  3.6.1+.  Additionally, AVI output mode requires "mencod-
er".  GIF output mode  requires  "gifsicle" 1.48+, "imagemagick" 6.3+,
and "intergif" 6.15+.

License:  This program is licensed under Creative Commons Attribution-
NonCommercial-ShareAlike 3.0.

END
        exit ONE;
    }

#---------------------------------------------------------------------
# If necessary, print usage text and exit.

    if ($cmderr || ($run == FALSE))
    {
        $DOCUMENTATION =~ s@^\s+@@s;
        $DOCUMENTATION =~ s@(__META_REVISION__)\n(-)@$1 $2@;
        $DOCUMENTATION =~ s@__META_NAME__@$PROGNAME@g;
        $DOCUMENTATION =~ s@__META_REVISION__@$REVISION@g;
        $DOCUMENTATION =~ s@\s*\z@\n@s;

        print "\n", $DOCUMENTATION, "\n";
        exit ONE;
    }

#---------------------------------------------------------------------
# Check external programs.

                                # Build list of programs needed
    my @progs = ();
       @progs = qw (mencoder) if $is_avi;
       @progs = qw (convert gifsicle intergif) unless $is_avi;
    push (@progs, 'povray');

    for my $prog (@progs)       # Check all  programs
    {                           # Check next program
        $str =  `which "$prog" 2>&1`;
        $str =  "" unless defined $str;
        $str =~ s@\n.*\z@@s;
                                # Error if program is missing
        die "Error: Operation needs the program $prog\n"
            unless (-f $str) && (-x $str);
    }

#---------------------------------------------------------------------
# Check for rest of package.

                                # Build list of remaining files
    my @files = qw (albedo_simp750.jpg moonphasepic.pl);

    for my $file (@files)       # Check all  files
    {                           # Check next file
        next if -f $file;       # If file exists, skip to next one
                                # File list as a string
        $str = join ' ', @files;

        print STDERR << "END";  # Print error message
Error: This program should be run in the directory that contains:
$str
END
        exit ONE;               # Error exit
    }

#---------------------------------------------------------------------
# Additional setup.

                                # Move some of the parameters
    my $FPS       = $is_avi ? $AVI_FPS       : $GIF_FPS         ;
    my $NUMFRAMES = $is_avi ? $AVI_NUMFRAMES : $GIF_NUMFRAMES   ;
    my $OFNAME    = $is_avi ? $AVI_OFNAME    : $GIF_OFNAME      ;
    my $TFNAME    = $is_avi ? $AVI_TFNAME    : $GIF_TFNAME      ;
    my $WIDTH     = $is_avi ? $AVI_WIDTH     : $GIF_WIDTH       ;

                                # Compute image height (in pixels)
    my $HEIGHT = int (($WIDTH * 0.75) + 0.5);

                                # Check "write" access
    die "Error: Can't write to current directory: $!\n"
        unless open (OFD, ">$TFNAME") &&
            close (OFD) && unlink ($TFNAME);

                                # Remove old files
    system "rm -f moonphase*.{gif,png} $OFNAME";

                                # Consistency check
    die "Error: Couldn't remove old file ($OFNAME)\n"
        if (-e $OFNAME) || (-l $OFNAME);

                                # Max. frame number (starting at zero)
    my $MaxFrameNum = $NUMFRAMES - ONE;

#---------------------------------------------------------------------
# Main loop.

    for my $ii (ZERO..$MaxFrameNum)
    {                           # Generate the next frame
        my $nn = sprintf ('%03d', $ii);
        my $xx = sprintf ('%.3f', $ii/$NUMFRAMES);

        my $pngfile = "moonphase$nn.png";
        my $giffile = "moonphase$nn.gif";

        system << "END";        # "END" must be double-quoted here
perl moonphasepic.pl $xx --output=$pngfile --width=$WIDTH
END
        if (!$is_avi)           # Is the target an  animated GIF ?
        {                       # Yes - Convert PNG frame to GIF frame
            system "convert $pngfile $giffile";
            unlink $pngfile;    # Discard PNG frame
        }
    }

#---------------------------------------------------------------------
# If GIF mode is selected, make an animated GIF.

# The  "gifsicle" command executed below creates a temporary (unoptim-
# ized)  animated-GIF  file  using 256 colors.  The "intergif" command
# executed afterwards  creates the final (optimized) animated-GIF file
# using $IGNUMCOLORS colors.

# Note: The extra spaces used in the $cmd block defined below are sig-
# nificant.  Don't  add  spaces or  remove them  unless you know  what
# you're doing.

    if (!$is_avi)               # Is the target an animated GIF ?
    {                           # Yes
        my $IGDELAY;
        $IGDELAY = int ((($FPS / $NUMFRAMES) * 100) + 0.5);
        $IGDELAY = 2 if $IGDELAY < 2;

        $cmd = << "END";        # "END" must be double-quoted here
gifsicle -i -O -l --colors 256
    moonphase[0-9][0-9][0-9].gif -o $TFNAME
intergif -i -t -trim -d $IGDELAY -loop -diffuse -zigzag
    -best $IGNUMCOLORS $TFNAME -o $OFNAME
rm -f $TFNAME moonphase[0-9][0-9][0-9].gif
END
        $cmd =~ s@\s*\n[\011\040]+@ @gs;
        $cmd =~ s@\s+\z@@s;
        system $cmd;
    }

#---------------------------------------------------------------------
# If AVI mode is selected, make an AVI file.

# Note: The extra spaces used in the $cmd block defined below are sig-
# nificant.  Don't  add  spaces or  remove them  unless you know  what
# you're doing.

    if ($is_avi)                # Is the target an AVI file?
    {                           # Yes
        $cmd = << "END";        # "END" must be double-quoted here
mencoder "mf://moonphase*.png"
    -mf w=$WIDTH:h=$HEIGHT:fps=$FPS:type=png
    -ovc lavc -lavcopts vcodec=mpeg4 -oac copy -o $OFNAME
rm -f moonphase*.png
END
        $cmd =~ s@\s*\n[\011\040]+@ @gs;
        $cmd =~ s@\s+\z@@s;
        system $cmd;
    }

#---------------------------------------------------------------------
# Wrap it up.

    die "Error: Operation failed\n" unless -f $OFNAME;
    print "Created $OFNAME\n";
    undef;
}

#---------------------------------------------------------------------
#                            main program
#---------------------------------------------------------------------

&Main();                        # Call the main routine
exit ZERO;                      # Normal exit
